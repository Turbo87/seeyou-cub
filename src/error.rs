use crate::Point;

pub type Result<T> = std::result::Result<T, Error>;

/// Unrecoverable parsing errors
#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error(transparent)]
    IoError(#[from] std::io::Error),

    #[error("Invalid magic bytes in header (expected 0x425543C2)")]
    InvalidMagicBytes,

    #[error("Encrypted CUB files not supported (encryption format undocumented)")]
    EncryptedFile,

    #[error("Invalid header offset: {found}")]
    InvalidHeaderOffset { found: i32 },

    #[error("Unexpected point flag: 0x{0:02X}")]
    UnexpectedPointFlag(u8),

    #[error("SizeOfItem is smaller than the minimum structure size")]
    UndersizedItems { size_of_item: i32 },

    #[error("SizeOfPoint is smaller than the minimum structure size")]
    UndersizedPoints { size_of_point: i32 },

    #[error("Coordinate out of valid range (lat: {}, lon: {})", .point.lat, .point.lon)]
    CoordinateOutOfRange { point: Point },
}
