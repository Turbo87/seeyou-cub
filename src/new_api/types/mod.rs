//! Type definitions for new API

// Re-export existing types that we'll reuse
pub use crate::types::{AltStyle, CubClass, CubStyle, DaysActive, ExtendedType};
pub use crate::types::{Header, NotamCodes, NotamScope, NotamTraffic, NotamType};
