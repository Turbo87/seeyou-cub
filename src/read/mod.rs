mod header;
mod io;
mod item;
mod item_data;

pub use header::parse_header;
pub use item::ItemIterator;
pub use item_data::read_item_data;
