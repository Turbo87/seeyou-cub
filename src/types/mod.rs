mod enums;
mod header;
mod item;
mod point;

pub use enums::*;
pub use header::*;
pub use item::*;
pub use point::*;

/// Parsed CUB file with header, items, and reader for lazy point parsing
pub struct CubFile<R> {
    header: Header,
    items: Vec<Item>,
    reader: R,
}

impl<R> CubFile<R> {
    /// Create new CubFile (used internally by parser)
    pub(crate) fn new(header: Header, items: Vec<Item>, reader: R) -> Self {
        Self {
            header,
            items,
            reader,
        }
    }

    /// Get file header
    pub fn header(&self) -> &Header {
        &self.header
    }

    /// Get all airspace items
    pub fn items(&self) -> &[Item] {
        &self.items
    }
}
