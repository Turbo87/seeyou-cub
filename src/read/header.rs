use crate::error::{Error, Result, Warning};
use crate::read::io::*;
use crate::types::{ByteOrder, Header};
use std::io::{Read, Seek, SeekFrom};

/// Parse CUB header (first 210 bytes)
pub fn parse_header<R: Read + Seek>(reader: &mut R) -> Result<(Header, Vec<Warning>)> {
    let mut warnings = Vec::new();

    // Seek to start
    reader.seek(SeekFrom::Start(0))?;

    // Read and validate magic bytes (offset 0-3, always LE)
    let ident = {
        let mut buf = [0u8; 4];
        reader.read_exact(&mut buf)?;
        u32::from_le_bytes(buf)
    };

    if ident != 0x425543C2 {
        return Err(Error::InvalidMagicBytes);
    }

    // Read title (offset 4-115, 112 bytes)
    let title = read_string(reader, 112)?.trim_end_matches('\0').to_string();

    // Read allowed serials (offset 116-131, 8 Ã— u16)
    // Parse as LE initially, will re-interpret if needed after reading PcByteOrder
    let allowed_serials = {
        let mut serials = [0u16; 8];
        for serial in &mut serials {
            *serial = read_u16(reader, ByteOrder::LE)?;
        }
        serials
    };

    // Read PcByteOrder (offset 132)
    let pc_byte_order = read_u8(reader)?;
    let byte_order = ByteOrder::from_pc_byte_order(pc_byte_order);

    // Re-read allowed_serials if byte order is BE
    let allowed_serials = if byte_order == ByteOrder::BE {
        reader.seek(SeekFrom::Start(116))?;
        let mut serials = [0u16; 8];
        for serial in &mut serials {
            *serial = read_u16(reader, ByteOrder::BE)?;
        }
        reader.seek(SeekFrom::Start(133))?; // Skip back to after PcByteOrder
        serials
    } else {
        allowed_serials
    };

    // Read IsSecured (offset 133)
    let is_secured = read_u8(reader)?;

    // Check encryption
    if is_secured != 0 {
        return Err(Error::EncryptedFile);
    }

    // Read Crc32 (offset 134-137)
    let crc32 = read_u32(reader, byte_order)?;

    // Read Key (offset 138-153, 16 bytes)
    let key = {
        let bytes = read_bytes(reader, 16)?;
        let mut key = [0u8; 16];
        key.copy_from_slice(&bytes);
        key
    };

    // Read remaining header fields (all use determined byte_order)
    let size_of_item = read_i32(reader, byte_order)?;
    let size_of_point = read_i32(reader, byte_order)?;
    let hdr_items = read_i32(reader, byte_order)?;
    let max_pts = read_i32(reader, byte_order)?;

    // Floats are always LE
    let left = read_f32_le(reader)?;
    let top = read_f32_le(reader)?;
    let right = read_f32_le(reader)?;
    let bottom = read_f32_le(reader)?;
    let max_width = read_f32_le(reader)?;
    let max_height = read_f32_le(reader)?;
    let lo_la_scale = read_f32_le(reader)?;

    let header_offset = read_i32(reader, byte_order)?;
    let data_offset = read_i32(reader, byte_order)?;
    let alignment = read_i32(reader, byte_order)?;

    // Validate sizes
    if size_of_item < 42 {
        warnings.push(Warning::OversizedItem {
            expected: size_of_item,
            actual: 42,
        });
    }

    if size_of_point < 5 {
        warnings.push(Warning::OversizedItem {
            expected: size_of_point,
            actual: 5,
        });
    }

    let header = Header {
        title,
        allowed_serials,
        pc_byte_order,
        is_secured,
        crc32,
        key,
        size_of_item,
        size_of_point,
        hdr_items,
        max_pts,
        left,
        top,
        right,
        bottom,
        max_width,
        max_height,
        lo_la_scale,
        header_offset,
        data_offset,
        alignment,
    };

    Ok((header, warnings))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    fn minimal_header_bytes(byte_order: ByteOrder, encrypted: bool) -> Vec<u8> {
        let mut bytes = vec![0u8; 210];

        // Magic bytes (LE)
        bytes[0..4].copy_from_slice(&0x425543C2u32.to_le_bytes());

        // Title (offset 4)
        bytes[4..15].copy_from_slice(b"Test Header");

        // Allowed serials (offset 116, skip)

        // PcByteOrder (offset 132)
        bytes[132] = match byte_order {
            ByteOrder::BE => 0,
            ByteOrder::LE => 1,
        };

        // IsSecured (offset 133)
        bytes[133] = if encrypted { 1 } else { 0 };

        // Write minimal valid values for remaining fields
        let write_i32 = |bytes: &mut [u8], offset: usize, value: i32| {
            let val_bytes = match byte_order {
                ByteOrder::LE => value.to_le_bytes(),
                ByteOrder::BE => value.to_be_bytes(),
            };
            bytes[offset..offset + 4].copy_from_slice(&val_bytes);
        };

        write_i32(&mut bytes, 154, 42); // size_of_item
        write_i32(&mut bytes, 158, 5); // size_of_point
        write_i32(&mut bytes, 162, 0); // hdr_items
        write_i32(&mut bytes, 166, 100); // max_pts

        // Floats (always LE)
        bytes[170..174].copy_from_slice(&0.0f32.to_le_bytes()); // left
        bytes[174..178].copy_from_slice(&1.0f32.to_le_bytes()); // top
        bytes[178..182].copy_from_slice(&1.0f32.to_le_bytes()); // right
        bytes[182..186].copy_from_slice(&0.0f32.to_le_bytes()); // bottom
        bytes[186..190].copy_from_slice(&1.0f32.to_le_bytes()); // max_width
        bytes[190..194].copy_from_slice(&1.0f32.to_le_bytes()); // max_height
        bytes[194..198].copy_from_slice(&1.0f32.to_le_bytes()); // lo_la_scale

        write_i32(&mut bytes, 198, 210); // header_offset
        write_i32(&mut bytes, 202, 210); // data_offset
        write_i32(&mut bytes, 206, 0); // alignment

        bytes
    }

    #[test]
    fn parse_valid_le_header() {
        let bytes = minimal_header_bytes(ByteOrder::LE, false);
        let mut cursor = Cursor::new(bytes);
        let (header, warnings) = parse_header(&mut cursor).unwrap();

        assert_eq!(header.byte_order(), ByteOrder::LE);
        assert!(!header.is_encrypted());
        assert_eq!(header.size_of_item, 42);
        assert!(warnings.is_empty());
    }

    #[test]
    fn parse_valid_be_header() {
        let bytes = minimal_header_bytes(ByteOrder::BE, false);
        let mut cursor = Cursor::new(bytes);
        let (header, warnings) = parse_header(&mut cursor).unwrap();

        assert_eq!(header.byte_order(), ByteOrder::BE);
        assert_eq!(header.size_of_item, 42);
        assert!(warnings.is_empty());
    }

    #[test]
    fn invalid_magic_bytes() {
        let mut bytes = minimal_header_bytes(ByteOrder::LE, false);
        bytes[0] = 0xFF; // Corrupt magic
        let mut cursor = Cursor::new(bytes);

        match parse_header(&mut cursor) {
            Err(Error::InvalidMagicBytes) => {}
            _ => panic!("Expected InvalidMagicBytes error"),
        }
    }

    #[test]
    fn encrypted_file_error() {
        let bytes = minimal_header_bytes(ByteOrder::LE, true);
        let mut cursor = Cursor::new(bytes);

        match parse_header(&mut cursor) {
            Err(Error::EncryptedFile) => {}
            _ => panic!("Expected EncryptedFile error"),
        }
    }
}
