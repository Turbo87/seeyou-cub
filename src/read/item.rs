use crate::error::{Result, Warning};
use crate::read::io::*;
use crate::types::{Header, Item};
use std::io::{Read, Seek, SeekFrom};

/// Parse all items from CUB file
pub fn parse_items<R: Read + Seek>(
    reader: &mut R,
    header: &Header,
) -> Result<(Vec<Item>, Vec<Warning>)> {
    let warnings = Vec::new();
    let byte_order = header.byte_order();

    // Seek to items section
    reader.seek(SeekFrom::Start(header.header_offset as u64))?;

    let mut items = Vec::with_capacity(header.hdr_items as usize);

    for _ in 0..header.hdr_items {
        // Create 43-byte zero-filled buffer and read SizeOfItem bytes into it
        // Per spec: remaining bytes should be set to 0 if SizeOfItem < 43
        let mut item_buffer = [0u8; 43];
        let bytes_to_read = std::cmp::min(header.size_of_item as usize, 43);
        reader.read_exact(&mut item_buffer[..bytes_to_read])?;

        // If SizeOfItem > 43, skip the extra bytes
        if header.size_of_item > 43 {
            skip_bytes(reader, (header.size_of_item - 43) as usize)?;
        }

        // Parse from the buffer using a cursor (full 43-byte buffer, zero-padded if needed)
        let mut cursor = std::io::Cursor::new(&item_buffer);

        let left = read_f32_le(&mut cursor)?;
        let top = read_f32_le(&mut cursor)?;
        let right = read_f32_le(&mut cursor)?;
        let bottom = read_f32_le(&mut cursor)?;

        let type_byte = read_u8(&mut cursor)?;
        let alt_style_byte = read_u8(&mut cursor)?;
        let min_alt = read_i16(&mut cursor, byte_order)?;
        let max_alt = read_i16(&mut cursor, byte_order)?;
        let points_offset = read_i32(&mut cursor, byte_order)?;
        let time_out = read_i32(&mut cursor, byte_order)?;
        let extra_data = read_u32(&mut cursor, byte_order)?;
        let active_time = read_u64(&mut cursor, byte_order)?;
        let extended_type_byte = read_u8(&mut cursor)?;

        items.push(Item {
            left,
            top,
            right,
            bottom,
            type_byte,
            alt_style_byte,
            min_alt,
            max_alt,
            points_offset,
            time_out,
            extra_data,
            active_time,
            extended_type_byte,
        });
    }

    Ok((items, warnings))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::ByteOrder;
    use std::io::Cursor;

    fn minimal_header() -> Header {
        Header {
            title: String::new(),
            allowed_serials: [0; 8],
            pc_byte_order: 1,
            is_secured: 0,
            crc32: 0,
            key: [0; 16],
            size_of_item: 43,
            size_of_point: 5,
            hdr_items: 2,
            max_pts: 100,
            left: 0.0,
            top: 1.0,
            right: 1.0,
            bottom: 0.0,
            max_width: 1.0,
            max_height: 1.0,
            lo_la_scale: 1.0,
            header_offset: 0,
            data_offset: 86, // 2 items Ã— 43 bytes
            alignment: 0,
        }
    }

    fn build_item_bytes(byte_order: ByteOrder) -> Vec<u8> {
        let mut bytes = Vec::new();

        // Item 1
        bytes.extend_from_slice(&0.1f32.to_le_bytes()); // left
        bytes.extend_from_slice(&0.5f32.to_le_bytes()); // top
        bytes.extend_from_slice(&0.4f32.to_le_bytes()); // right
        bytes.extend_from_slice(&0.2f32.to_le_bytes()); // bottom

        bytes.push(0x04); // type (DA)
        bytes.push(0x23); // alt_style (max=FL, min=MSL)

        let write_i16 = |bytes: &mut Vec<u8>, val: i16| {
            bytes.extend_from_slice(&match byte_order {
                ByteOrder::LE => val.to_le_bytes(),
                ByteOrder::BE => val.to_be_bytes(),
            });
        };
        let write_i32 = |bytes: &mut Vec<u8>, val: i32| {
            bytes.extend_from_slice(&match byte_order {
                ByteOrder::LE => val.to_le_bytes(),
                ByteOrder::BE => val.to_be_bytes(),
            });
        };
        let write_u32 = |bytes: &mut Vec<u8>, val: u32| {
            bytes.extend_from_slice(&match byte_order {
                ByteOrder::LE => val.to_le_bytes(),
                ByteOrder::BE => val.to_be_bytes(),
            });
        };
        let write_u64 = |bytes: &mut Vec<u8>, val: u64| {
            bytes.extend_from_slice(&match byte_order {
                ByteOrder::LE => val.to_le_bytes(),
                ByteOrder::BE => val.to_be_bytes(),
            });
        };

        write_i16(&mut bytes, 100); // min_alt
        write_i16(&mut bytes, 5000); // max_alt
        write_i32(&mut bytes, 0); // points_offset
        write_i32(&mut bytes, 0); // time_out
        write_u32(&mut bytes, 0); // extra_data
        write_u64(&mut bytes, 0x3FFFFFF); // active_time (default)
        bytes.push(0); // extended_type

        // Item 2 (copy of item 1 for simplicity)
        let item1 = bytes.clone();
        bytes.extend_from_slice(&item1);

        bytes
    }

    #[test]
    fn parse_items_le() {
        let header = minimal_header();
        let bytes = build_item_bytes(ByteOrder::LE);
        let mut cursor = Cursor::new(bytes);

        let (items, warnings) = parse_items(&mut cursor, &header).unwrap();

        assert_eq!(items.len(), 2);
        assert_eq!(items[0].left, 0.1);
        assert_eq!(items[0].min_alt, 100);
        assert_eq!(items[0].max_alt, 5000);
        assert!(warnings.is_empty());
    }
}
