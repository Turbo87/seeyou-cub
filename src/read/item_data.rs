use crate::error::{Error, Result, Warning};
use crate::read::io::*;
use crate::types::{CubDataId, Header, Item, ItemData, Point};
use std::io::{Read, Seek, SeekFrom};

/// Parse complete item data (geometry + metadata) for an airspace item
pub fn read_item_data<R: Read + Seek>(
    reader: &mut R,
    header: &Header,
    item: &Item,
    warnings: &mut Vec<Warning>,
) -> Result<ItemData> {
    // Calculate offset for first point
    let points_offset = header.data_offset as u64 + item.points_offset as u64;
    reader.seek(SeekFrom::Start(points_offset))?;

    // Initialize origin to item's bottom-left
    let mut origin_x = item.left;
    let mut origin_y = item.bottom;

    let byte_order = header.byte_order();

    let mut points = Vec::new();

    // Phase 1: Parse geometry points
    loop {
        let flag = match read_u8(reader) {
            Ok(flag) => flag,
            Err(Error::IoError(ref e)) if e.kind() == std::io::ErrorKind::UnexpectedEof => {
                // Hit EOF - no attributes section
                return Ok(ItemData {
                    points,
                    name: None,
                    frequency: None,
                    frequency_name: None,
                    icao_code: None,
                    secondary_frequency: None,
                    exception_rules: None,
                    notam_remarks: None,
                    notam_id: None,
                    notam_insert_time: None,
                });
            }
            Err(e) => return Err(e),
        };

        match flag {
            0x81 => {
                // Origin update
                let delta_x = read_i16(reader, byte_order)? as f32 * header.lo_la_scale;
                let delta_y = read_i16(reader, byte_order)? as f32 * header.lo_la_scale;
                origin_x += delta_x;
                origin_y += delta_y;
            }

            0x01 => {
                // Geometry point
                let x = read_i16(reader, byte_order)? as f32 * header.lo_la_scale;
                let y = read_i16(reader, byte_order)? as f32 * header.lo_la_scale;

                let lon = origin_x + x;
                let lat = origin_y + y;

                points.push(Point { lon, lat });
            }

            flag if (flag & 0x40) != 0 => {
                // Attribute section - parse and return
                return parse_attributes(reader, header, flag, warnings, points);
            }

            _ => {
                return Err(Error::UnexpectedPointFlag(flag));
            }
        }
    }
}

struct OptionalFields {
    icao_code: Option<String>,
    secondary_frequency: Option<u32>,
    exception_rules: Option<String>,
    notam_remarks: Option<String>,
    notam_id: Option<String>,
    notam_insert_time: Option<u32>,
}

impl OptionalFields {
    fn new() -> Self {
        Self {
            icao_code: None,
            secondary_frequency: None,
            exception_rules: None,
            notam_remarks: None,
            notam_id: None,
            notam_insert_time: None,
        }
    }
}

/// Parse attribute section starting with given flag
fn parse_attributes<R: Read + Seek>(
    reader: &mut R,
    header: &Header,
    first_flag: u8,
    warnings: &mut Vec<Warning>,
    points: Vec<Point>,
) -> Result<ItemData> {
    let byte_order = header.byte_order();

    let mut name = None;
    let mut frequency = None;
    let mut frequency_name = None;
    let mut optional = OptionalFields::new();

    // First attribute: name
    if (first_flag & 0x40) != 0 {
        // Skip remaining bytes of point structure
        skip_bytes(reader, (header.size_of_point - 1) as usize)?;

        let name_len = (first_flag & 0x3F) as usize;
        if name_len > 0 {
            name = Some(read_string(reader, name_len)?);
        }
    }

    // Check for frequency attribute
    let next_flag = match read_u8(reader) {
        Ok(flag) => flag,
        Err(Error::IoError(ref e)) if e.kind() == std::io::ErrorKind::UnexpectedEof => {
            // EOF after name - no frequency or optional data
            return Ok(ItemData {
                points,
                name,
                frequency,
                frequency_name,
                icao_code: optional.icao_code,
                secondary_frequency: optional.secondary_frequency,
                exception_rules: optional.exception_rules,
                notam_remarks: optional.notam_remarks,
                notam_id: optional.notam_id,
                notam_insert_time: optional.notam_insert_time,
            });
        }
        Err(e) => return Err(e),
    };

    if (next_flag & 0xC0) == 0xC0 {
        // Frequency attribute
        let freq_name_len = (next_flag & 0x3F) as usize;
        frequency = Some(read_u32(reader, byte_order)?);

        if freq_name_len > 0 {
            frequency_name = Some(read_string(reader, freq_name_len)?);
        }

        // Read next flag for optional data
        parse_optional_data(reader, warnings, &mut optional)?;
    } else if next_flag == 0xA0 {
        // Optional data without frequency
        parse_optional_data_with_flag(reader, next_flag, warnings, &mut optional)?;
    }
    // If neither frequency nor optional data, we're done (flag not recognized, ignore)

    Ok(ItemData {
        points,
        name,
        frequency,
        frequency_name,
        icao_code: optional.icao_code,
        secondary_frequency: optional.secondary_frequency,
        exception_rules: optional.exception_rules,
        notam_remarks: optional.notam_remarks,
        notam_id: optional.notam_id,
        notam_insert_time: optional.notam_insert_time,
    })
}

/// Parse optional data records (0xA0 flags)
fn parse_optional_data<R: Read + Seek>(
    reader: &mut R,
    warnings: &mut Vec<Warning>,
    optional: &mut OptionalFields,
) -> Result<()> {
    loop {
        let flag = match read_u8(reader) {
            Ok(flag) => flag,
            Err(Error::IoError(ref e)) if e.kind() == std::io::ErrorKind::UnexpectedEof => {
                return Ok(());
            }
            Err(e) => return Err(e),
        };

        if flag != 0xA0 {
            // Not optional data, we're done
            return Ok(());
        }

        parse_optional_data_record(reader, warnings, optional)?;
    }
}

/// Parse optional data with flag already read
fn parse_optional_data_with_flag<R: Read + Seek>(
    reader: &mut R,
    flag: u8,
    warnings: &mut Vec<Warning>,
    optional: &mut OptionalFields,
) -> Result<()> {
    if flag == 0xA0 {
        parse_optional_data_record(reader, warnings, optional)?;
        parse_optional_data(reader, warnings, optional)?;
    }
    Ok(())
}

/// Parse single optional data record
fn parse_optional_data_record<R: Read>(
    reader: &mut R,
    warnings: &mut Vec<Warning>,
    optional: &mut OptionalFields,
) -> Result<()> {
    let data_id = read_u8(reader)?;
    let b1 = read_u8(reader)?;
    let b2 = read_u8(reader)?;
    let b3 = read_u8(reader)?;

    let data_id_enum = CubDataId::from_byte(data_id);

    match data_id_enum {
        Some(CubDataId::IcaoCode) => {
            let len = b3 as usize;
            optional.icao_code = Some(read_string(reader, len)?);
        }

        Some(CubDataId::SecondaryFrequency) => {
            let value = ((b1 as u32) << 16) | ((b2 as u32) << 8) | (b3 as u32);
            optional.secondary_frequency = Some(value);
        }

        Some(CubDataId::ExceptionRules) => {
            let len = (((b2 as u16) << 8) | (b3 as u16)) as usize;
            optional.exception_rules = Some(read_string(reader, len)?);
        }

        Some(CubDataId::NotamRemarks) => {
            let len = (((b2 as u16) << 8) | (b3 as u16)) as usize;
            optional.notam_remarks = Some(read_string(reader, len)?);
        }

        Some(CubDataId::NotamId) => {
            let len = b3 as usize;
            optional.notam_id = Some(read_string(reader, len)?);
        }

        Some(CubDataId::NotamInsertTime) => {
            let b4 = read_u8(reader)?;
            let value =
                ((b1 as u32) << 24) | ((b2 as u32) << 16) | ((b3 as u32) << 8) | (b4 as u32);
            optional.notam_insert_time = Some(value);
        }

        None => {
            warnings.push(Warning::UnknownPointFlag(data_id));
        }
    }

    Ok(())
}
