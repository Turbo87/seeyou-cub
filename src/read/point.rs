use std::io::{Read, Seek, SeekFrom};
use crate::error::{Result, Warning};
use crate::types::{Header, Item, ParsedPoint, OptionalData, CubDataId};
use crate::read::io::*;

/// Iterator that lazily parses CubPoint sequences for an item
pub struct PointIterator<'a, R> {
    reader: &'a mut R,
    header: &'a Header,
    item: Item,
    origin_x: f32,
    origin_y: f32,
    done: bool,
    warnings: Vec<Warning>,
    // Attributes parsed from current point sequence
    current_name: Option<String>,
    current_frequency: Option<u32>,
    current_frequency_name: Option<String>,
    current_optional: Vec<OptionalData>,
}

impl<'a, R: Read + Seek> PointIterator<'a, R> {
    /// Create new point iterator for an item
    pub(crate) fn new(
        reader: &'a mut R,
        header: &'a Header,
        item: &Item,
    ) -> Result<Self> {
        // Seek to first point for this item
        let offset = header.data_offset as u64 + item.points_offset as u64;
        reader.seek(SeekFrom::Start(offset))?;

        // Initialize origin to item's bottom-left
        let origin_x = item.left;
        let origin_y = item.bottom;

        Ok(Self {
            reader,
            header,
            item: item.clone(),
            origin_x,
            origin_y,
            done: false,
            warnings: Vec::new(),
            current_name: None,
            current_frequency: None,
            current_frequency_name: None,
            current_optional: Vec::new(),
        })
    }

    /// Get warnings collected during parsing
    pub fn warnings(&self) -> &[Warning] {
        &self.warnings
    }

    /// Parse next CubPoint
    fn parse_next_point(&mut self) -> Result<Option<ParsedPoint>> {
        if self.done {
            return Ok(None);
        }

        let byte_order = self.header.byte_order();

        loop {
            let flag = match read_u8(self.reader) {
                Ok(flag) => flag,
                Err(crate::error::Error::IoError(ref e)) if e.kind() == std::io::ErrorKind::UnexpectedEof => {
                    // Hit EOF while trying to read next flag - treat as end of points
                    self.done = true;
                    return Ok(None);
                },
                Err(e) => return Err(e),
            };

            match flag {
                0x81 => {
                    // Set origin offset
                    let delta_x = read_i16(self.reader, byte_order)? as f32 * self.header.lo_la_scale;
                    let delta_y = read_i16(self.reader, byte_order)? as f32 * self.header.lo_la_scale;
                    self.origin_x += delta_x;
                    self.origin_y += delta_y;

                    // Continue reading
                }

                0x01 => {
                    // New point
                    let x = read_i16(self.reader, byte_order)? as f32 * self.header.lo_la_scale;
                    let y = read_i16(self.reader, byte_order)? as f32 * self.header.lo_la_scale;

                    let lon = self.origin_x + x;
                    let lat = self.origin_y + y;

                    // Build point with current attributes
                    let point = ParsedPoint {
                        lon,
                        lat,
                        name: self.current_name.take(),
                        frequency: self.current_frequency.take(),
                        frequency_name: self.current_frequency_name.take(),
                        optional_data: std::mem::take(&mut self.current_optional),
                    };

                    return Ok(Some(point));
                }

                flag if (flag & 0x40) != 0 => {
                    // Attribute block
                    self.parse_attributes(flag)?;
                }

                0x00 => {
                    // End of points
                    self.done = true;
                    return Ok(None);
                }

                _ => {
                    // Unknown flag - collect warning and skip
                    self.warnings.push(Warning::UnknownPointFlag(flag));
                    // Try to skip this point (size_of_point - 1 for flag already read)
                    skip_bytes(self.reader, (self.header.size_of_point - 1) as usize)?;
                }
            }
        }
    }

    /// Parse attribute records starting with given flag
    fn parse_attributes(&mut self, first_flag: u8) -> Result<()> {
        let byte_order = self.header.byte_order();

        // First attribute: name
        if (first_flag & 0x40) != 0 {
            let name_len = (first_flag & 0x3F) as usize;
            if name_len > 0 {
                let name = read_string(self.reader, name_len)?;
                self.current_name = Some(name.trim_end_matches('\0').to_string());
            }
        }

        // Check for frequency attribute
        let next_flag = read_u8(self.reader)?;
        if (next_flag & 0xC0) == 0xC0 {
            let freq_name_len = (next_flag & 0x3F) as usize;
            let frequency = read_u32(self.reader, byte_order)?;
            self.current_frequency = Some(frequency);

            if freq_name_len > 0 {
                let freq_name = read_string(self.reader, freq_name_len)?;
                self.current_frequency_name = Some(freq_name.trim_end_matches('\0').to_string());
            }

            // Read next flag for optional data
            self.parse_optional_data()?;
        } else if next_flag == 0xA0 {
            // Optional data without frequency
            self.parse_optional_data_with_flag(next_flag)?;
        } else {
            // No more attributes, seek back one byte
            use std::io::SeekFrom;
            self.reader.seek(SeekFrom::Current(-1))?;
        }

        Ok(())
    }

    /// Parse optional data records (0xA0 flags)
    fn parse_optional_data(&mut self) -> Result<()> {
        loop {
            let flag = read_u8(self.reader)?;
            if flag != 0xA0 {
                // Not optional data, seek back
                use std::io::SeekFrom;
                self.reader.seek(SeekFrom::Current(-1))?;
                break;
            }

            self.parse_optional_data_record()?;
        }

        Ok(())
    }

    /// Parse optional data with flag already read
    fn parse_optional_data_with_flag(&mut self, flag: u8) -> Result<()> {
        if flag == 0xA0 {
            self.parse_optional_data_record()?;
            self.parse_optional_data()?;
        }
        Ok(())
    }

    /// Parse single optional data record
    fn parse_optional_data_record(&mut self) -> Result<()> {
        let data_id = read_u8(self.reader)?;
        let b1 = read_u8(self.reader)?;
        let b2 = read_u8(self.reader)?;
        let b3 = read_u8(self.reader)?;

        let data_id_enum = CubDataId::from_byte(data_id);

        match data_id_enum {
            Some(CubDataId::IcaoCode) => {
                let len = b3 as usize;
                let icao = read_string(self.reader, len)?;
                self.current_optional.push(OptionalData::IcaoCode(icao));
            }

            Some(CubDataId::SecondaryFrequency) => {
                let value = ((b1 as u32) << 16) | ((b2 as u32) << 8) | (b3 as u32);
                self.current_optional.push(OptionalData::SecondaryFrequency(value));
            }

            Some(CubDataId::ExceptionRules) => {
                let len = (((b2 as u16) << 8) | (b3 as u16)) as usize;
                let rules = read_string(self.reader, len)?;
                self.current_optional.push(OptionalData::ExceptionRules(rules));
            }

            Some(CubDataId::NotamRemarks) => {
                let len = (((b2 as u16) << 8) | (b3 as u16)) as usize;
                let remarks = read_string(self.reader, len)?;
                self.current_optional.push(OptionalData::NotamRemarks(remarks));
            }

            Some(CubDataId::NotamId) => {
                let len = b3 as usize;
                let id = read_string(self.reader, len)?;
                self.current_optional.push(OptionalData::NotamId(id));
            }

            Some(CubDataId::NotamInsertTime) => {
                let b4 = read_u8(self.reader)?;
                let value = ((b1 as u32) << 16) | ((b2 as u32) << 8) | (b3 as u32);
                let time = (value << 8) | (b4 as u32);
                self.current_optional.push(OptionalData::NotamInsertTime(time));
            }

            None => {
                self.warnings.push(Warning::UnknownPointFlag(data_id));
            }
        }

        Ok(())
    }
}

impl<'a, R: Read + Seek> Iterator for PointIterator<'a, R> {
    type Item = Result<ParsedPoint>;

    fn next(&mut self) -> Option<Self::Item> {
        self.parse_next_point().transpose()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    fn minimal_header() -> Header {
        Header {
            title: String::new(),
            allowed_serials: [0; 8],
            pc_byte_order: 1,
            is_secured: 0,
            crc32: 0,
            key: [0; 16],
            size_of_item: 42,
            size_of_point: 5,
            hdr_items: 1,
            max_pts: 100,
            left: 0.0,
            top: 1.0,
            right: 1.0,
            bottom: 0.0,
            max_width: 1.0,
            max_height: 1.0,
            lo_la_scale: 0.0001,
            header_offset: 0,
            data_offset: 0,  // Points start at beginning for this test
            alignment: 0,
        }
    }

    fn minimal_item() -> Item {
        Item {
            left: 0.0,
            top: 1.0,
            right: 1.0,
            bottom: 0.0,
            type_byte: 0x04,
            alt_style_byte: 0,
            min_alt: 0,
            max_alt: 1000,
            points_offset: 0,
            time_out: 0,
            extra_data: 0,
            active_time: 0,
            extended_type_byte: 0,
        }
    }

    #[test]
    fn parse_simple_point() {
        let mut bytes = Vec::new();

        // Point: 0x01 flag + coords
        bytes.push(0x01);
        bytes.extend_from_slice(&100i16.to_le_bytes());  // x offset
        bytes.extend_from_slice(&200i16.to_le_bytes());  // y offset

        // End marker
        bytes.push(0x00);

        let mut cursor = Cursor::new(bytes);
        let header = minimal_header();
        let item = minimal_item();

        let mut iter = PointIterator::new(&mut cursor, &header, &item).unwrap();

        let point = iter.next().unwrap().unwrap();
        assert!((point.lon - (0.0 + 100.0 * 0.0001)).abs() < 0.00001);
        assert!((point.lat - (0.0 + 200.0 * 0.0001)).abs() < 0.00001);
        assert!(point.name.is_none());

        assert!(iter.next().is_none());
    }
}