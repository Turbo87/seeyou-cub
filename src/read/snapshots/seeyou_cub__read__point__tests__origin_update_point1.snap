---
source: src/read/point.rs
expression: point1
---
ParsedPoint {
    lon: 0.105,
    lat: 0.21,
    name: None,
    frequency: None,
    frequency_name: None,
    optional_data: [],
}
